File: platforms/ios/Providers/UniTrackFirebase/Sources/FirebaseMessagingHelper.swift
// FirebaseMessagingHelper.swift
//
// Helpers the app calls from its existing UNUserNotificationCenterDelegate /
// MessagingDelegate methods to fan out push-notification + FCM-token events
// into UniTrack (and from there to Snowplow + the portal).
//
// We deliberately don't take over the delegate: every app worth migrating
// already owns those delegates (FPT Life keeps the FCM token in Keychain,
// the notification routing logic, etc.). Forcing a swizzle-based proxy here
// would conflict with that. Instead the app keeps its delegate and adds one
// line per callback.
//
// Usage in AppDelegate / FSSAppDelegate:
//
//   func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
//       UniTrackFirebaseMessaging.handleTokenUpdate(fcmToken)
//   }
//
//   func userNotificationCenter(_ center: UNUserNotificationCenter,
//                                didReceive response: UNNotificationResponse,
//                                withCompletionHandler completionHandler: @escaping () -> Void) {
//       UniTrackFirebaseMessaging.handleNotificationClicked(response)
//       completionHandler()
//   }
//
// When FirebaseMessaging isn't linked into the app (e.g. a non-push build),
// the whole file compiles down to no-op stubs so call sites stay portable.

import Foundation
import UniTrack
import UserNotifications

#if canImport(FirebaseMessaging)
import FirebaseMessaging

public enum UniTrackFirebaseMessaging {

    // Cache the latest token so identify() / setUser can attach it as a trait
    // without the app having to plumb the value through itself.
    private static let lock = NSLock()
    private static var cachedToken: String?

    /// Latest FCM token seen by the SDK (read-only). nil until the first
    /// token-update callback fires.
    public static var currentToken: String? {
        lock.lock(); defer { lock.unlock() }
        return cachedToken
    }

    /// Call from `messaging(_:didReceiveRegistrationToken:)`. Fires
    /// `fcm_token_updated` whenever the token actually changes (Firebase will
    /// echo the same token on every cold start — we dedupe to avoid noise).
    public static func handleTokenUpdate(_ token: String?) {
        lock.lock()
        let prev = cachedToken
        cachedToken = token
        lock.unlock()

        guard let token = token, !token.isEmpty, token != prev else { return }
        var props: [String: Any] = ["fcm_token": token]
        if let p = prev, !p.isEmpty { props["prev_token"] = p }
        UniTrack.track("fcm_token_updated", properties: props)
    }

    /// Call from `userNotificationCenter(_:didReceive:withCompletionHandler:)`.
    /// Treated as the moment the user actually engaged with the push.
    public static func handleNotificationClicked(_ response: UNNotificationResponse) {
        let n = response.notification
        let userInfo = n.request.content.userInfo
        let (notifId, title, body, data) = extract(notification: n, userInfo: userInfo)
        UniTrack.trackNotification(
            state: "foreground",
            action: "clicked",
            title: title,
            body: body,
            notificationId: notifId,
            data: data
        )
    }

    /// Call from `userNotificationCenter(_:willPresent:withCompletionHandler:)`.
    /// Fires whenever a push arrives while the app is in the foreground (iOS
    /// asks the delegate whether to display it).
    public static func handleNotificationReceivedForeground(_ notification: UNNotification) {
        let userInfo = notification.request.content.userInfo
        let (notifId, title, body, data) = extract(notification: notification, userInfo: userInfo)
        UniTrack.trackNotification(
            state: "foreground",
            action: "received",
            title: title,
            body: body,
            notificationId: notifId,
            data: data
        )
    }

    /// Call from `application(_:didReceiveRemoteNotification:fetchCompletionHandler:)`
    /// when the app is in background / silent push. No UNNotification object
    /// available there, so we work from the raw userInfo bag.
    public static func handleSilentPush(_ userInfo: [AnyHashable: Any]) {
        let notifId = userInfo["gcm.message_id"] as? String
            ?? userInfo["google.message_id"] as? String
            ?? ""
        let (title, body) = extractAlert(userInfo: userInfo)
        UniTrack.trackNotification(
            state: "background",
            action: "received",
            title: title,
            body: body,
            notificationId: notifId,
            data: jsonSafe(userInfo)
        )
    }

    // MARK: - Internal extraction helpers

    /// Pull the bits we want out of an iOS notification + its userInfo. We
    /// prefer fields off UNNotification when present (already decoded by iOS)
    /// and fall back to APS/data dictionaries.
    private static func extract(notification: UNNotification,
                                 userInfo: [AnyHashable: Any]) -> (String, String?, String?, [String: Any]) {
        let id = (userInfo["gcm.message_id"] as? String)
            ?? (userInfo["google.message_id"] as? String)
            ?? notification.request.identifier
        let content = notification.request.content
        let title: String? = content.title.isEmpty ? nil : content.title
        let body:  String? = content.body.isEmpty  ? nil : content.body
        // Use the iOS-decoded title/body when set; otherwise the helper below
        // tries to parse aps.alert directly.
        if title == nil && body == nil {
            let (t, b) = extractAlert(userInfo: userInfo)
            return (id, t, b, jsonSafe(userInfo))
        }
        return (id, title, body, jsonSafe(userInfo))
    }

    /// APS alert can be either a string (legacy) or a dict ({title, body}).
    private static func extractAlert(userInfo: [AnyHashable: Any]) -> (String?, String?) {
        guard let aps = userInfo["aps"] as? [AnyHashable: Any] else { return (nil, nil) }
        if let alertStr = aps["alert"] as? String {
            return (nil, alertStr)
        }
        if let alert = aps["alert"] as? [AnyHashable: Any] {
            return (alert["title"] as? String, alert["body"] as? String)
        }
        return (nil, nil)
    }

    /// Strip non-JSON-safe values + coerce keys to String. APNs payloads
    /// occasionally include NSNumber / NSDate / nested AnyHashable keys; the
    /// portal's JSON encoder rejects those.
    private static func jsonSafe(_ bag: [AnyHashable: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for (k, v) in bag {
            let key = String(describing: k)
            if let s = v as? String       { out[key] = s; continue }
            if let n = v as? NSNumber     { out[key] = n; continue }
            if let b = v as? Bool         { out[key] = b; continue }
            if let arr = v as? [Any]      { out[key] = arr; continue }
            if let sub = v as? [AnyHashable: Any] { out[key] = jsonSafe(sub); continue }
            out[key] = String(describing: v)
        }
        return out
    }
}

#else

public enum UniTrackFirebaseMessaging {
    public static var currentToken: String? { nil }
    public static func handleTokenUpdate(_ token: String?) {}
    public static func handleNotificationClicked(_ response: UNNotificationResponse) {}
    public static func handleNotificationReceivedForeground(_ notification: UNNotification) {}
    public static func handleSilentPush(_ userInfo: [AnyHashable: Any]) {}
}

#endif

File: platforms/flutter/unitrack_firebase/lib/firebase_messaging_helper.dart
// UniTrackFirebaseMessaging — push/notification + FCM token tracking helpers.
//
// Apps already own their FirebaseMessaging callbacks (token rotation handler,
// onMessage / onMessageOpenedApp listeners). This helper does NOT take over
// those callbacks — instead the app calls one method per callback site and
// gets:
//
//   • fcm_token_updated event (deduped against previous token)
//   • notification event with state=foreground/background/silent + action
//     received/clicked, routed through UniTrack.trackNotification so the
//     portal + Snowplow + Firebase Analytics all see the same notification.
//
// Wire-up (typical app):
//
//   FirebaseMessaging.instance.onTokenRefresh.listen(
//     UniTrackFirebaseMessaging.handleTokenUpdate);
//
//   FirebaseMessaging.onMessage.listen(
//     UniTrackFirebaseMessaging.handleNotificationReceivedForeground);
//
//   FirebaseMessaging.onMessageOpenedApp.listen(
//     UniTrackFirebaseMessaging.handleNotificationClicked);
//
//   final initialMsg = await FirebaseMessaging.instance.getInitialMessage();
//   if (initialMsg != null) {
//     UniTrackFirebaseMessaging.handleNotificationClicked(initialMsg);
//   }

import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:unitrack/unitrack.dart';

class UniTrackFirebaseMessaging {
  UniTrackFirebaseMessaging._();

  static String? _cachedToken;

  /// Latest FCM token seen by the helper (null until [handleTokenUpdate]
  /// has been called at least once).
  static String? get currentToken => _cachedToken;

  /// Wire to `FirebaseMessaging.instance.onTokenRefresh` AND the value from
  /// `getToken()` at startup. Fires `fcm_token_updated` only when the token
  /// actually changes — Firebase re-emits the same token on every cold start
  /// and we don't want to spam analytics with duplicates.
  static void handleTokenUpdate(String? token) {
    final prev = _cachedToken;
    _cachedToken = token;
    if (token == null || token.isEmpty || token == prev) return;
    final props = <String, Object?>{'fcm_token': token};
    if (prev != null && prev.isNotEmpty) props['prev_token'] = prev;
    UniTrack.instance.track('fcm_token_updated', properties: props);
  }

  /// Wire to `FirebaseMessaging.onMessage`. Fires when a push lands while the
  /// app is in the foreground (iOS calls willPresent → app gets a chance to
  /// display; Android delivers to onMessage directly).
  static void handleNotificationReceivedForeground(RemoteMessage message) {
    UniTrack.instance.trackNotification(
      state: 'foreground',
      action: 'received',
      title: message.notification?.title,
      body:  message.notification?.body,
      notificationId: _idOf(message),
      data: _safeData(message.data),
    );
  }

  /// Wire to `FirebaseMessaging.onMessageOpenedApp` AND
  /// `FirebaseMessaging.instance.getInitialMessage()` (which fires when the
  /// app launched from a tapped push). Treated as the moment the user
  /// actually engaged.
  static void handleNotificationClicked(RemoteMessage message) {
    UniTrack.instance.trackNotification(
      state: 'foreground',
      action: 'clicked',
      title: message.notification?.title,
      body:  message.notification?.body,
      notificationId: _idOf(message),
      data: _safeData(message.data),
    );
  }

  /// Wire to your background message handler (the `@pragma('vm:entry-point')`
  /// top-level function passed to `FirebaseMessaging.onBackgroundMessage`).
  /// Treated as a silent / background-arrival event.
  ///
  /// Note: background handlers run in an isolated Dart isolate; the UniTrack
  /// instance there has NO active session unless the app initializes it
  /// inside the entry-point. The event will still queue locally and flush
  /// once the main isolate revives.
  static void handleBackgroundMessage(RemoteMessage message) {
    final hasVisible = message.notification != null;
    UniTrack.instance.trackNotification(
      state: hasVisible ? 'background' : 'silent',
      action: 'received',
      title: message.notification?.title,
      body:  message.notification?.body,
      notificationId: _idOf(message),
      data: _safeData(message.data),
    );
  }

  // ── helpers ────────────────────────────────────────────────────────────

  // FCM messageId is the canonical id when present; on iOS the same id is
  // mirrored under `gcm.message_id` / `google.message_id` in the data bag.
  // Fall back to those for parity with the Swift implementation.
  static String? _idOf(RemoteMessage m) {
    if (m.messageId != null && m.messageId!.isNotEmpty) return m.messageId;
    final d = m.data;
    final g = d['gcm.message_id'] ?? d['google.message_id'];
    if (g is String && g.isNotEmpty) return g;
    return null;
  }

  // RemoteMessage.data is already Map<String,String>, but be defensive in
  // case future Firebase versions widen the value type.
  static Map<String, Object?> _safeData(Map<String, dynamic> data) {
    if (data.isEmpty) return const <String, Object?>{};
    final out = <String, Object?>{};
    data.forEach((k, v) {
      if (v is String || v is num || v is bool || v == null) out[k] = v;
      else out[k] = v.toString();
    });
    return out;
  }
}
